syntax = "proto3";

import "google/protobuf/struct.proto";

option java_multiple_files = true;
option java_package = "org.hopenly.ilyde.grpc";
option java_outer_classname = "DatasetProto";
option objc_class_prefix = "DS";

package datasets;

// The datasets service definition.
service DatasetServices {
    // Retrieve a dataset With ObjectID
    rpc RetrieveDataset (ID) returns (Dataset) {}
    //  Create a dataset using information
    rpc CreateDataset (Dataset) returns (Dataset) {}
    // Update a dataset using information
    rpc UpdateDataset (Dataset) returns (Dataset) {}
    // Delete a dataset with ObjectID
    rpc DeleteDataset (ID) returns (OperationStatus) {}
    // Search for datasets
    rpc SearchDatasets (SearchDatasetRequest) returns (SearchDatasetResponse) {}
    // Retrieve a dataset's version passing the version'ID
    rpc RetrieveDatasetVersion (ID) returns (Version) {}
    //  Create a dataset version passing a Dataset's ID for which to create a version
    rpc CreateDatasetVersion (google.protobuf.Struct) returns (Version) {}
    // Search for dataset's versions
    rpc SearchDatasetVersions (SearchVersionRequest) returns (SearchVersionResponse) {}
    // Create bucket
    rpc CreateBucket (Bucket) returns (Bucket) {}
}

// message key/value pair
message Bucket {
    string name = 1;
}

// message key/value pair
message File {
    string name = 1;
    uint64 size = 2;
}

// The dataset message representing a dataset metadata
message Dataset {
    string id = 1;
    string name = 2;
    string description = 3;
    repeated google.protobuf.Struct labels = 4;
    string version = 5;
    string create_at = 6;
    string last_update = 7;
}

// The Version message representing a Dataset Version
message Version {
    string id = 1;
    string name = 2;
    string dataset = 3;
    string related_bucket = 4;
    repeated File bucket_tree = 5;
    uint64 size = 6;
    string create_at = 7;
}

// Object ID message
message ID {
    string id = 1;
}

// message OperationStatus
message OperationStatus {
    uint32 status = 1;
    string message = 2;
}

// message for Datasets filtering
message SearchDatasetRequest {
    google.protobuf.Struct filter = 1;
    uint32 page_number = 2;
    uint32 result_per_page = 3;
}

// search response for dataset
message  SearchDatasetResponse {
    uint32 total = 1; // search total count results
    uint32 page_number = 2;
    uint32 result_per_page = 3;
    repeated Dataset datasets = 4;
}

// message  searching versions of a dataset
message SearchVersionRequest {
    google.protobuf.Struct filter = 1;
    int32 page_number = 2;
    int32 result_per_page = 3;
}

// message for searching response for versions
message SearchVersionResponse {
    int32 total = 1; // search total count results
    int32 page_number = 2;
    int32 result_per_page = 3;
    repeated Version versions = 4;
}